reana_workflow_controller/rest.py

```diff
@@ -1861,3 +1995,91 @@ def _mark_workflow_as_deleted_in_db(workflow):
     current_db_sessions = Session.object_session(workflow)
     current_db_sessions.add(workflow)
     current_db_sessions.commit()
+
+
+def get_specification_diff(workflow_a, workflow_b, output_format=None):
+    """Return differences between two workflow specifications.
+
+    :param workflow_a: The first workflow to be compared.
+    :type: reana_db.models.Workflow instance.
+    :param workflow_a: The first workflow to be compared.
+    :type: reana_db.models.Workflow instance.
+    :param output_format: Sets output format. Optional.
+    :type: String. One of ['unified', 'context', 'html'].
+           Unified format returned if not set.
+
+    :rtype: List with lines of differences.
+    """
+    if output_format not in ['unified', 'context', 'html', None]:
+        raise ValueError('Unknown output format.'
+                         'Please select one of unified, context or html.')
+
+    if output_format in [None, 'unified']:
+        diff_method = getattr(difflib, 'unified_diff')
+        # specification_diff = list(difflib.unified_diff(spec_a, spec_b))[2:]
+        # specification_diff = [line.splitlines()[0] for line
+        #                       in specification_diff]
+    elif output_format == 'context':
+        diff_method = getattr(difflib, 'context_diff')
+        # specification_diff = list(difflib.context_diff(spec_a, spec_b))
+    elif output_format == 'html':
+        diff_method = getattr(difflib, 'HtmlDiff')
+        # specification_diff = difflib.HtmlDiff(spec_a, spec_b)
+    specification_diff = dict.fromkeys(workflow_a.reana_specification.keys())
+    for section in specification_diff:
+        section_a = pprint.pformat(
+            workflow_a.reana_specification.get(section, '')).\
+            splitlines()
+        section_b = pprint.pformat(
+            workflow_b.reana_specification.get(section, '')).\
+            splitlines()
+        specification_diff[section] = list(diff_method(section_a,
+                                                       section_b))[2:]
+    return specification_diff
+
+
+def get_workspace_diff(workflow_a, workflow_b, brief=False, context_lines=5):
+    """Return differences between two workspaces.
+
+    :param workflow_a: The first workflow to be compared.
+    :type: reana_db.models.Workflow instance.
+    :param workflow_b: The second workflow to be compared.
+    :type: reana_db.models.Workflow instance.
+    :param brief: Optional flag to show brief workspace diff.
+    :type: Boolean.
+    :param context_lines: The number of context lines to show above and after
+                          the discovered differences.
+    :type: Integer or string.
+
+    :rtype: Dictionary with file paths and their sizes
```

Comment:
This would be `:return` and `:rtype` dict, no? @reanahub/developers are we happy to start adding type annotations? Since here we are running on Python 3 on all cluster components I would rather start adding Python 3 type annotations than this ones.

Reply:
This follows the docstring styles in the utility methods in `rest.py`